src/Microsoft.Diagnostics.Monitoring.EventPipe/Triggers/Pipelines/EventPipeTriggerPipeline.cs

```diff
@@ -0,0 +1,95 @@
+        // once) over the lifetime of the pipeline, depending on the implementation of the trigger.
+        private readonly Action<TraceEvent> _callback;
+
+        /// <summary>
+        /// The pipeline used to monitor the trace event source from the event pipe using the trigger
+        /// specified in the settings of the current pipeline.
+        /// </summary>
+        private TraceEventTriggerPipeline _pipeline;
+
+        // The trigger implementation used to detect a condition in the trace event source.
+        private ITraceEventTrigger _trigger;
+
+        public EventPipeTriggerPipeline(DiagnosticsClient client, EventPipeTriggerPipelineSettings<TSettings> settings, Action<TraceEvent> callback) :
+            base(client, settings)
+        {
+            if (null == Settings.Configuration)
+            {
+                throw new ArgumentException(FormattableString.Invariant($"The {nameof(settings.Configuration)} property on the settings must not be null."), nameof(settings));
+            }
+
+            if (null == Settings.TriggerFactory)
+            {
+                throw new ArgumentException(FormattableString.Invariant($"The {nameof(settings.TriggerFactory)} property on the settings must not be null."), nameof(settings));
+            }
+
+            _callback = callback;
+        }
+
+        protected override MonitoringSourceConfiguration CreateConfiguration()
+        {
+            return Settings.Configuration;
+        }
+
+        protected override async Task OnEventSourceAvailable(EventPipeEventSource eventSource, Func<Task> stopSessionAsync, CancellationToken token)
+        {
+            _trigger = Settings.TriggerFactory.Create(Settings.TriggerSettings);
+
+            _pipeline = new TraceEventTriggerPipeline(eventSource, _trigger, _callback);
+
+            await _pipeline.RunAsync(token).ConfigureAwait(false);
+        }
+
+        protected override async Task OnStop(CancellationToken token)
+        {
+            if (null != _pipeline)
+            {
+                await _pipeline.StopAsync(token).ConfigureAwait(false);
+            }
+            await base.OnStop(token);
+        }
+
+        protected override async Task OnCleanup()
+        {
+            if (null != _pipeline)
+            {
+                await _pipeline.DisposeAsync().ConfigureAwait(false);
+            }
+
+            // Disposal is not part of the ITraceEventTrigger interface; check the implementation
+            // of the trigger to see if it implements one of the disposal interfaces and call it.
+            if (_trigger is IAsyncDisposable asyncDisposableTrigger)
+            {
+                await asyncDisposableTrigger.DisposeAsync().ConfigureAwait(false);
+            }
+            else if (_trigger is IDisposable disposableTrigger)
+            {
+                disposableTrigger.Dispose();
+            }
```

Comment:
Rather than guess, could you make it explicit on the interface and have a default implementation for `IDisposable`/`IAsyncDisposable` on the interface?

Reply:
I didn't want to force `IDisposable`/`IAsyncDisposable` in the contract. For example, the `EventCounterTrigger` doesn't need it since it is just storing timestamp/tick values. I wanted to allow that to be an opt-in behavior. Given that `EventPipeTriggerPipeline` is the class that created the trigger instance, I felt that giving the trigger an opportunity to dispose (without having to wait for destruction) without forcing disposal on the trigger was a good balance of capabilities.